File: wrap/csllbc/csharp/core/config/Ini.cs
// The MIT License (MIT)

// Copyright (c) 2013 lailongwei<lailongwei@126.com>
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of 
// this software and associated documentation files (the "Software"), to deal in 
// the Software without restriction, including without limitation the rights to 
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of 
// the Software, and to permit persons to whom the Software is furnished to do so, 
// subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all 
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS 
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;

namespace llbc
{
    #region IniSection
    /// <summary>
    /// The ini section class encapsulation.
    /// </summary>
    public class IniSection : IDisposable
    {
        #region Constructor/Destructor
        /// <summary>
        /// Create new Ini section object
        /// </summary>
        /// <param name="sectionName">section name</param>
        /// <param name="values">section values</param>
        public IniSection(string sectionName, Dictionary<string, string> values)
        {
            if ((_sectionName = sectionName) == null)
                throw new LLBCException("Section name could not be null");
            if ((_values = values) == null)
                throw new LLBCException("Section values could not be null");
        }

        ~IniSection()
        {
            Dispose(false);
        }
        #endregion

        #region sectionName, values
        /// <summary>
        /// Get section name.
        /// </summary>
        public string sectionName
        {
            get { return _sectionName; }
        }

        /// <summary>
        /// Get all section values
        /// </summary>
        public Dictionary<string, string> values
        {
            get { return _values; }
        }
        #endregion

        #region Subscript operator
        /// <summary>
        /// Get specific key's value, if not found, will raise exception.
        /// </summary>
        /// <param name="key">config key</param>
        /// <returns>config value</returns>
        public string this[string key]
        {
            get
            {
                if (!_values.ContainsKey(key))
                    throw new LLBCException("Not found value in IniSection, key:{0}", key);

                return _values[key];
            }
        }
        #endregion

        #region HasKey/GetValue<T>
        /// <summary>
        /// Check ini section exist specific key config or not.
        /// </summary>
        /// <param name="key">the config key</param>
        /// <returns>return true if has specific key config, otherwise return false</returns>
        public bool HasKey(string key)
        {
            return _values.ContainsKey(key);
        }

        /// <summary>
        /// Get specific key's value(template method), if not found, will return dftValue.
        /// </summary>
        /// <typeparam name="T">the except return value type</typeparam>
        /// <param name="key">the config key</param>
        /// <param name="dftValue">default config value</param>
        /// <returns></returns>
        public T GetValue<T>(string key, T dftValue = default(T))
        {
            string value;
            if (!_values.TryGetValue(key, out value))
                return dftValue;

            return StringUtil.ToType<T>(value, dftValue);
        }
        #endregion

        #region Dispose implement
        protected virtual void Dispose(bool disposing)
        {
            if (_disposed)
                return;

            _sectionName = null;
            _values = null;

            _disposed = true;
            if (disposing)
                GC.SuppressFinalize(this);
        }

        public void Dispose()
        {
            Dispose(true);
        }
        #endregion

        private bool _disposed;

        private string _sectionName;
        private Dictionary<string, string> _values;
    }
    #endregion

    #region Ini
    /// <summary>
    /// The ini class encapsulation, use for parse ini config format file/content.
    /// </summary>
    public class Ini : IDisposable
    {
        #region Constructor/Destructor
        public Ini()
        {
            _sections = new Dictionary<string, IniSection>();
        }

        ~Ini()
        {
            Dispose(false);
        }
        #endregion

        #region sectionNames, sections
        public List<string> sectionNames
        {
            get
            {
                return new List<string>(_sections.Keys);
            }
        }

        public Dictionary<string, IniSection> sections
        {
            get
            {
                return _sections;
            }
        }
        #endregion

        #region Subscript operator implementation
        /// <summary>
        /// Get specific section name's ini section object, if not found will raise exception.
        /// </summary>
        /// <param name="sectionName">the section name</param>
        /// <returns>specific name's ini section object</returns>
        public IniSection this[string sectionName]
        {
            get
            {
                var section = GetSection(sectionName);
                if (section == null)
                    throw new LLBCException("Not found section[{0}] in Ini object", sectionName);

                return section;
            }
        }
        #endregion

        #region LoadFromContent/LoadFromFile
        /// <summary>
        /// Load ini config data from ini format string content.
        /// </summary>
        /// <param name="content">ini format string content</param>
        public void LoadFromContent(string content)
        {
            _sections.Clear();

            unsafe
            {
                int sectionCount = 0;
                IntPtr nativeContent = LibUtil.CreateNativeStr(content);

                int errDescLen = 0;
                IntPtr nativeSections = 
                    LLBCNative.csllbc_Ini_LoadFromContent(nativeContent, new IntPtr(&sectionCount), new IntPtr(&errDescLen));
                LibUtil.FreeNativePtr(ref nativeContent, false);

                if (sectionCount == -1)
                {
                    byte* nativeErrorDesc = (byte*)nativeSections;
                    string errorDesc = LibUtil.Ptr2Str(nativeErrorDesc, errDescLen);
                    LibUtil.FreeNativePtr(nativeErrorDesc);

                    throw new LLBCException("{0}", errorDesc);
                }
                else if (sectionCount == 0)
                    return;

                _LoadAllSections((_NativeIniSection*)nativeSections.ToPointer(), sectionCount);
                LLBCNative.csllbc_Ini_FreeNativeSections(nativeSections, sectionCount);
            }
        }

        /// <summary>
        /// Load ini config data from ini config file.
        /// </summary>
        /// <param name="file">ini config file</param>
        public void LoadFromFile(string file)
        {
            byte[] bytes = null;
            using (var fs = new FileStream(file, FileMode.Open, FileAccess.Read))
            {
                if (fs.Length == 0)
                {
                    _sections.Clear();
                    return;
                }

                bytes = new byte[fs.Length];
                fs.Read(bytes, 0, (int)fs.Length);
            }

            LoadFromContent(Encoding.UTF8.GetString(bytes));
        }
        #endregion

        #region HasSection/GetSection, HasKey/GetValue
        /// <summary>
        /// Check has specific section name's config or not.
        /// </summary>
        /// <param name="sectionName">the section name</param>
        /// <returns>true if exist specific section name's config, otherwise return false</returns>
        public bool HasSection(string sectionName)
        {
            return _sections.ContainsKey(sectionName);
        }

        /// <summary>
        /// Get specific name's ini section object, if not found, return null(no exception raise).
        /// </summary>
        /// <param name="sectionName">the section name</param>
        /// <returns>the ini section object, if not found, return null</returns>
        public IniSection GetSection(string sectionName)
        {
            IniSection section = null;
            _sections.TryGetValue(sectionName, out section);
            return section;
        }

        /// <summary>
        /// Check has specific key's config or not.
        /// </summary>
        /// <param name="sectionName">the section name</param>
        /// <param name="key">the config key</param>
        /// <returns>true if exist specific config key's config, otherwise return false</returns>
        public bool HasKey(string sectionName, string key)
        {
            IniSection section;
            if (!_sections.TryGetValue(sectionName, out section))
                return false;

            return section.HasKey(key);
        }

        /// <summary>
        /// Get specific key's config value, if not found, return default value(no exception raise).
        /// </summary>
        /// <typeparam name="T">the value type</typeparam>
        /// <param name="sectionName">the section name</param>
        /// <param name="key">the config key</param>
        /// <param name="dftValue">if not found, will return this default value</param>
        /// <returns>the specific key's config value, if not found will return dftValue parameter specified value</returns>
        public T GetValue<T>(string sectionName, string key, T dftValue = default(T))
        {
            IniSection section = GetSection(sectionName);
            return section != null ? section.GetValue<T>(key, dftValue) : dftValue;
        }
        #endregion

        #region Dispose implement
        public void Dispose()
        {
            Dispose(true);
        }

        virtual protected void Dispose(bool disposing)
        {
            if (_sections == null)
                return;

            _sections = null;
            if (disposing)
                GC.SuppressFinalize(this);
        }
        #endregion

        #region Internal implementation
        private unsafe void _LoadAllSections(_NativeIniSection* nativeSections, int sectionCount)
        {
            for (int sectionIdx = 0; sectionIdx < sectionCount; ++sectionIdx)
            {
                _NativeIniSection nativeSection = *(nativeSections + sectionIdx);
                string sectionName = LibUtil.Ptr2Str(nativeSection.sectionName, nativeSection.sectionNameLen);

                Dictionary<string, string> values = new Dictionary<string, string>(nativeSection.count);
                for (int valueIdx = 0; valueIdx < nativeSection.count; ++valueIdx)
                {
                    string key = LibUtil.Ptr2Str(nativeSection.keys[valueIdx], nativeSection.keysLen[valueIdx]);
                    string value = LibUtil.Ptr2Str(nativeSection.values[valueIdx], nativeSection.valuesLen[valueIdx]);
                    values.Add(key, value);
                }

                _sections.Add(sectionName, new IniSection(sectionName, values));
            }
        }
        #endregion
        
        private Dictionary<string, IniSection> _sections;
    }
    #endregion

    #region Internal structure encapsulation.
    /// <summary>
    /// Library internal structure define, use for swap manage and unmanaged memory data.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    internal unsafe struct _NativeIniSection
    {
        public byte* sectionName;
        public int sectionNameLen;

        public byte** keys;
        public int* keysLen;
        public byte** values;
        public int* valuesLen;

        public int count;
    }
    #endregion
}

File: wrap/csllbc/native/src/core/config/_Ini.cpp
// The MIT License (MIT)

// Copyright (c) 2013 lailongwei<lailongwei@126.com>
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of 
// this software and associated documentation files (the "Software"), to deal in 
// the Software without restriction, including without limitation the rights to 
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of 
// the Software, and to permit persons to whom the Software is furnished to do so, 
// subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all 
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS 
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include "csllbc/common/Export.h"

#include "csllbc/core/config/_Ini.h"

namespace
{
    csllbc_Ini_IniSection *_GetAllSections(LLBC_Ini *ini, int *sectionCount)
    {
        const std::map<LLBC_String, LLBC_IniSection *> &sections = ini->GetAllSections();
        if ((*sectionCount = static_cast<int>(sections.size())) == 0)
            return nullptr;

        int sectionIdx = 0;
        csllbc_Ini_IniSection *nativeSections = LLBC_Malloc(csllbc_Ini_IniSection, sizeof(csllbc_Ini_IniSection) * *sectionCount);
        for (std::map<LLBC_String, LLBC_IniSection *>::const_iterator sectionIt = sections.begin();
             sectionIt != sections.end();
             sectionIt++)
        {
            // Get section name.
            const LLBC_String &sectionName = sectionIt->first;
            csllbc_Ini_IniSection *nativeSection = nativeSections + sectionIdx++;
            nativeSection->sectionName = 
                csllbc_PInvokeUtil::BuildCSharpStr(sectionName, &nativeSection->sectionNameLen);
            
            // Get all key-values pair in this section.
            int valueIdx = 0;
            const LLBC_IniSection *section = sectionIt->second;

            const std::map<LLBC_String, LLBC_Variant> &values = section->GetAllValues();

            size_t valueCount = values.size();
            nativeSection->count = static_cast<int>(values.size());
            if (valueCount == 0)
            {
                nativeSection->keys = nullptr;
                nativeSection->keysLen = nullptr;
                nativeSection->values = nullptr;
                nativeSection->valuesLen = nullptr;
                continue;
            }

            nativeSection->keys = LLBC_Malloc(char *, sizeof(char *) * valueCount);
            nativeSection->values = LLBC_Malloc(char *, sizeof(char *) * valueCount);
            nativeSection->keysLen = LLBC_Malloc(int, sizeof(int) * valueCount);
            nativeSection->valuesLen = LLBC_Malloc(int, sizeof(int) * valueCount);
            for (std::map<LLBC_String, LLBC_Variant>::const_iterator valueIt = values.begin();
                 valueIt != values.end();
                 valueIt++)
            {
                char *&nativeKey = nativeSection->keys[valueIdx];
                int &nativeKeyLen = nativeSection->keysLen[valueIdx];

                char *&nativeValue = nativeSection->values[valueIdx];
                int &nativeValueLen = nativeSection->valuesLen[valueIdx];

                valueIdx++;

                const LLBC_String &key = valueIt->first;
                nativeKey = csllbc_PInvokeUtil::BuildCSharpStr(key, &nativeKeyLen);
                const LLBC_Variant &value = valueIt->second;
                nativeValue = csllbc_PInvokeUtil::BuildCSharpStr(value, &nativeValueLen);
            }
        }

        return nativeSections;
    }
}

LLBC_BEGIN_C_DECL

csllbc_Ini_IniSection *csllbc_Ini_LoadFromContent(const char *content, int *sectionCount, int *errDescLen)
{
    LLBC_Ini *ini = new LLBC_Ini;
    if (ini->LoadFromContent(content) != LLBC_OK)
    {
        LLBC_String errStr = LLBC_String().format(
            "load ini failed, error: %s, detail information: %s",
            LLBC_FormatLastError(), ini->GetLoadError().c_str());

        *sectionCount = -1;
        return reinterpret_cast<csllbc_Ini_IniSection *>(csllbc_PInvokeUtil::BuildCSharpStr(errStr, errDescLen));
    }

    csllbc_Ini_IniSection *sections = _GetAllSections(ini, sectionCount);
    delete ini;

    return sections;
}

void csllbc_Ini_FreeNativeSections(csllbc_Ini_IniSection *sections, int sectionCount)
{
    if (UNLIKELY(sections == nullptr))
        return;

    for (int sectionIdx = 0; sectionIdx < sectionCount; ++sectionIdx)
    {
        csllbc_Ini_IniSection &section = sections[sectionIdx];
        free(section.sectionName);

        int valueCount = section.count;
        for (int valueIdx = 0; valueIdx < valueCount; ++valueIdx)
        {
            free(section.keys[valueIdx]);
            LLBC_XFree(section.values[valueIdx]);
        }

        if (valueCount > 0)
        {
            free(section.keys);
            free(section.keysLen);

            free(section.values);
            free(section.valuesLen);
        }
    }

    free(sections);
}

LLBC_END_C_DECL